### PR TITLE
CT-8 Step CRUD

### PR DESCRIPTION
**Pull request by brama051**

### CODE REVIEW

File: packages/web/template.env [hunk 1]
```diff
@@ -3,7 +3,7 @@ REACT_APP_URL="localhost:3007"
 REACT_APP_API_URL="/api"
 REACT_APP_SOCKET_URL="/api/socket.io"
 
-REACT_APP_SENTRY_DSN="DSN URL"
+REACT_APP_SENTRY_DSN="https://fa4694414d9249bf9c5534034f702ede@o1079031.ingest.sentry.io/6083593"
```
Comment (MrBr) [inline packages/web/template.env]: Rebasing to the development should solve this issue. Aleks updated frontend so it doesn't crash without sentry dns.

File: packages/models/src/step/service.ts [hunk 2]
```diff
@@ -428,6 +442,11 @@ const createStep = <T>(
   id: string,
   stepType: T extends Step<infer U, infer K> ? K : never,
   state: T extends Step<infer U, infer K> ? U : never,
+  owner?: User,
```
Comment (MrBr) [inline packages/models/src/step/service.ts]: I'd rather have last argument like `data` or something like that containing all the remaining data.

```
createStep = (id, stepType, state, data? = {}) => ({id, stepType, state, ...data})
```

File: packages/models/src/step/service.ts [hunk 3]
```diff
@@ -21,6 +24,17 @@ const isSameStep = <T extends {}>(
   return leftStep?.id === rightStep?.id;
 };
 
+/**
+ * Public copy of the lesson, accessible by non collaborators.
+ */
+const isStepPublicDocument = (step: Step) => !!step.docId;
```
Comment (MrBr) [inline packages/models/src/step/service.ts]: I don't think we need `docId` anymore. The way system is going to work looks a lot different than it used to. Doc id was helpful when everything revolved around lessons/template. Since this is not the case anymore I think we can drop it and find a better way for handling published subjects.

File: packages/server/src/modules/step/middleware.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+import * as service from './service';
+
+export const saveStep: MiddlewareFunction = (req, res, next) => {
+  service
+    .saveStep(res.locals.step as Step)
+    .then(next)
+    .catch(next);
+};
+
+export const publishStep: MiddlewareFunction = (req, res, next) => {
```
Comment (MrBr) [inline packages/server/src/modules/step/middleware.ts]: Following the comment above, let's drop logic related publishing.

File: packages/server/src/modules/step/middleware.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    .then(steps => {
+      if (!steps) {
+        throw new StepNotFoundError();
+      }
+      res.locals.steps = steps;
+      next();
+    })
+    .catch(next);
+};
+export const canEditStep: MiddlewareFunction = (req, res, next) => {
```
Comment (MrBr) [inline packages/server/src/modules/step/middleware.ts]: I'd drop this one as well. We're not going to have steps without the new permission. Sine it's not going to be deployed anytime soon we can have routes unchecked.

File: packages/server/src/modules/step/middleware.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      if (canEdit) {
+        next();
+        return;
+      }
+      throw new UnauthorizedStepEditError();
+    })
+    .catch(next);
+};
+
+export const canAccessStep: MiddlewareFunction = (req, res, next) => {
```
Comment (MrBr) [inline packages/server/src/modules/step/middleware.ts]: Same as the one above.

File: packages/server/src/modules/step/model.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
+import { Difficulty, NormalizedStep, TYPE_STEP, TYPE_TAG, TYPE_USER } from '@chess-tent/models';
+import { db } from '@application';
+import { Schema } from 'mongoose';
+
+// could be called StepEntity as entities are established concepts in the persistence context
+export interface DepopulatedStep {
+  id: NormalizedStep['id'];
+  docId: NormalizedStep['docId'];
```
Comment (MrBr) [inline packages/server/src/modules/step/model.ts]: Not needed.

File: packages/server/src/modules/step/model.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  type: NormalizedStep['type'];
+  owner: NormalizedStep['owner'];
+  state: NormalizedStep['state'];
+  difficulty?: NormalizedStep['difficulty'];
+  tags?: NormalizedStep['tags'];
+  published?: NormalizedStep['published'];
```
Comment (MrBr) [inline packages/server/src/modules/step/model.ts]: Not needed.

File: packages/server/src/modules/step/model.ts [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+  users?: NormalizedStep['users'];
```
Comment (MrBr) [inline packages/server/src/modules/step/model.ts]: Not needed, will be done thorough permissions.

File: packages/server/src/modules/step/model.ts [hunk 10]
```diff
@@ -0,0 +1,76 @@
+import { Difficulty, NormalizedStep, TYPE_STEP, TYPE_TAG, TYPE_USER } from '@chess-tent/models';
+import { db } from '@application';
+import { Schema } from 'mongoose';
+
+// could be called StepEntity as entities are established concepts in the persistence context
+export interface DepopulatedStep {
+  id: NormalizedStep['id'];
+  docId: NormalizedStep['docId'];
+  type: NormalizedStep['type'];
+  owner: NormalizedStep['owner'];
```
Comment (MrBr) [inline packages/server/src/modules/step/model.ts]: Will be done through the permissions?

File: packages/models/src/step/service.ts [hunk 11]
```diff
@@ -428,6 +442,11 @@ const createStep = <T>(
   id: string,
   stepType: T extends Step<infer U, infer K> ? K : never,
   state: T extends Step<infer U, infer K> ? U : never,
+  owner?: User,
+  users?: User[],
```
Comment (MrBr) [inline packages/models/src/step/service.ts]: Also, users and owner aren't probably needed since it's going to be done through the new permissions?

File: packages/server/src/modules/step/model.ts [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+  difficulty?: NormalizedStep['difficulty'];
+  tags?: NormalizedStep['tags'];
+  published?: NormalizedStep['published'];
+  users?: NormalizedStep['users'];
+  v?: number;
+}
+
+const stepSchema = db.createSchema<DepopulatedStep>(
+  {
+    docId: {
```
Comment (MrBr) [inline packages/server/src/modules/step/model.ts]: Reflect suggestions on the schema as well.

File: packages/server/src/modules/step/model.ts [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+      type: Schema.Types.Number,
+      default: 1,
+    } as unknown as number,
+  },
+  { minimize: false },
+);
+
+stepSchema.index({ 'state.id': 'text' });
+
+// do we need adapter before schema needs to be modified?
```
Comment (MrBr) [inline packages/server/src/modules/step/model.ts]: No, remove it. Adapter pattern here is going to garbage.

File: packages/server/src/modules/step/routes.ts [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+  findSteps,
+  patchStep,
+  canAccessStep,
+  unpublishStep,
+  deleteStep,
+} from './middleware';
+
+const { identify, sendData, sendStatusOk, toLocals } = middleware;
+
+application.service.registerPostRoute(
```
Comment (MrBr) [inline packages/server/src/modules/step/routes.ts:12]: Reflect aforementioned changes here as well.

File: packages/server/src/modules/step/service.ts [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+        }
+        resolve(result ? result.toObject<Step>() : null);
+      });
+  });
+
+export const deleteStep = async (stepId: Step['id']) => {
+  await StepModel.deleteOne({ _id: stepId });
+};
+
+export const publishStep = (stepId: Step['id']) =>
```
Comment (MrBr) [inline packages/server/src/modules/step/service.ts]: Kill it with fire.

File: packages/server/src/modules/step/service.ts [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+      err => {
+        if (err) {
+          throw err;
+        }
+        resolve();
+      },
+    );
+  });
+
+export const unpublishStep = (stepId: Step['id']) =>
```
Comment (MrBr) [inline packages/server/src/modules/step/service.ts]: Kill it with ice.

File: packages/models/src/lesson/types.ts [hunk 17]
```diff
@@ -6,6 +6,10 @@ import { Step } from '../step';
 import { Activity } from '../activity';
 
 export const TYPE_LESSON = 'lessons';
+export enum TYPE_SUBJECT {
```
Comment (MrBr) [inline packages/models/src/lesson/types.ts]: I think you can remove this one, maybe in `activity` model it makes sense to extend subject type?

**Review by MrBr — COMMENTED** [review]
Aside the comments related to the code I'd like to use this PR to establish BE testing practice.

Please spare some time thinking about what would you like to tests.

In my mind it makes sense to test individual service (units) and their integration through endpoint (integration) tests.

File: packages/models/src/lesson/types.ts (hunk 17 above)
Comment (brama051) [reply packages/models/src/lesson/types.ts]: That was my intention but we'll get to that.

File: packages/server/src/modules/step/middleware.ts (hunk 4 above)
Comment (brama051) [reply packages/server/src/modules/step/middleware.ts]: 👍

File: packages/server/src/modules/step/middleware.ts (hunk 6 above)
Comment (brama051) [reply packages/server/src/modules/step/middleware.ts]: 👍

File: packages/server/src/modules/step/model.ts (hunk 8 above)
Comment (brama051) [reply packages/server/src/modules/step/model.ts]: 👍

File: packages/server/src/modules/step/service.ts [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+        published: filters.published,
+        ...db.orQueries(owner, users),
+      });
+
+    if (filters.difficulty) {
+      query['difficulty'] = { $eq: filters.difficulty };
+    }
+
+    if (!_.isEmpty(filters.tagIds)) {
+      // query['tags'] = { $in: filters.tagIds }; // todo: compiler is barking about thisxx
```
Comment (brama051) [inline packages/server/src/modules/step/service.ts]: ???

Comment (MrBr) [reply packages/server/src/modules/step/service.ts]: `NormalizedStep` doesn't have normalized `tags` field.

https://github.com/MrBr/chess-tent/pull/9/files/76e1e1a18417908df3640882458450f1f387b944#diff-1a70e3e6f97eb97d154e67f2c441402082d593b85138695cb97419493f9d2765R27

```
  tags: Step['tags'][]; // should be Tag['id'][];
```

Comment (brama051) [reply packages/server/src/modules/step/service.ts]: Barking is no more!

File: packages/server/src/modules/step/__tests__/get-step-route.spec.ts [hunk 19]
```diff
@@ -0,0 +1,14 @@
+import request from 'supertest';
+
+import application from '@application';
+
+describe('GET /step/:stepId', () => {
```
Comment (MrBr) [inline packages/server/src/modules/step/__tests__/get-step-route.spec.ts:136]: Let's use this opportunity to do the rest of the setup.

- [ ] Define a way for mocking a token
- [ ] Create DB container for testing
- [ ] Create a separate env file for tests

@brama051 do you have anything else that was blocking you to add more tests?

File: packages/server/src/modules/app/index.ts [hunk 20]
```diff
@@ -82,7 +83,11 @@ application.start = async () => {
...
 
 application.test.start = async () => {
-  connect();
+  generateUniqueDbName();
+  await connect();
   // TODO - should be a part of the lifecycle hook/event
   app.use(application.middleware.errorHandler);
 };
+
+const generateUniqueDbName = () =>
```
Comment (MrBr) [inline packages/server/src/modules/app/index.ts]: `utils.ts` file in the same module is a better place for the function definition. Ideally trying to avoid index file pollution.

File: packages/server/src/application/tests/mongo.spec.ts [hunk 21]
```diff
@@ -1,47 +1,12 @@
...
-  Network,
-  StartedNetwork,
-  StartedTestContainer,
-  Wait,
-} from 'testcontainers';
-import * as process from 'process';
 import application from '@application';
 
 describe('Sample Test', () => {
-  let network: StartedNetwork;
```
Comment (MrBr) [inline packages/server/src/application/tests/mongo.spec.ts]: Nice! I like solutions where code is being deleted. Looks like we've made bootstrapping much more easier.

File: packages/server/src/modules/app/index.ts (hunk 20 above)
Comment (brama051) [reply packages/server/src/modules/app/index.ts]: Sure. This is at first step of "Make It Work, Make It Right, Make It Fast"

**Review by MrBr — DISMISSED** [review]

File: packages/server/src/modules/tag/defaults.ts [hunk 22]
```diff
@@ -14,9 +14,11 @@ const defaultTags = [
   'Mate',
 ].map(text => ({ text, id: service.generateIndex() }));
 
-TagModel.findOne({ text: defaultTags[0].text }).then(result => {
-  if (result) {
-    return;
-  }
-  TagModel.insertMany(defaultTags);
-});
+console.warn('Move to migration');
```
Comment (MrBr) [inline packages/server/src/modules/tag/defaults.ts]: You can undo this changes to avoid conflicts. I've moved it to migrations in the open PR.

**Review by MrBr — APPROVED** [review]
